### DISCUSSION TITLE
Legacy JAVA application with splash screen

### DISCUSSION THREAD

**Discussion by J-Bu** [udevbe/greenfield#141] [Q&A] [answered]
Hi,

first a huge thanks for creating the project it is very interesting to play with it.

I try to run a legacy JAVA application inside a docker container and use greenfield to render the UI. In general it works really great but I have a problem. The application uses a splashscreen and after the splash show the real window. It seems like the application tracking with XWayland does not work as expected because it terminates the app when the splashscreen is closed maybe because no active window is available for a short amount of time.
I digged into the source and hacked my way around it but this is definitely not the right way to do it. Maybe you can give me hint how the application tracking is done and where I could look to correctly fix it.

Here is the hack I use at the moment:
```
--- a/packages/compositor-proxy/src/NativeWaylandClientSession.ts
+++ b/packages/compositor-proxy/src/NativeWaylandClientSession.ts
@@ -77,10 +77,11 @@ export function createNativeClientSession(
   )
 
   setClientDestroyedCallback(wlClient, () => {
+    logger.info(`setClientDestroyedCallback`)
     for (const destroyListener of nativeClientSession.destroyListeners) {
       destroyListener()
     }
-    nativeAppContext.sendClientConnectionsDisconnect()
+    //nativeAppContext.sendClientConnectionsDisconnect()
     nativeClientSession.destroyListeners = []
     nativeClientSession.destroyed = true
   })
```

Regards,
Jan

**Comment by Zubnix** [comment]
Thanks for reporting this. I assume the splash is simply a configurable swing/awt splash screen? I'll investigate and fix 👍🏻

**Comment by J-Bu** [comment]
In case it helps I created a repo to generate the docker containers I use the sofware itself must be manually downloaded with requires to fill out a form with some information.

https://github.com/J-Bu/ums-docker

**Comment by J-Bu** [comment]
I tried a bit more it this seems only to happen when falling back to the software renderer. If I forward /dev/dri into the container I'm not able to reproduce the error. As soon as I remove the forward I can reproduce it again.

**Reply by Zubnix** [reply]
I'm not sure that is the issue/fix. I've written a simple hello-world swing app with a splash screen and run it on my dev- setup (including dri & hw accell) and it works without issues ~50% of the time, the other 45% the splash screen is shown but not the main window, the other 5% neither are shown. I suspect a weird race condition is the cause, I'll investigate further.

**Reply by Zubnix** [reply]
> Here is the hack I use at the moment:
> 
> ```
> --- a/packages/compositor-proxy/src/NativeWaylandClientSession.ts
> +++ b/packages/compositor-proxy/src/NativeWaylandClientSession.ts
> @@ -77,10 +77,11 @@ export function createNativeClientSession(
>    )
>  
>    setClientDestroyedCallback(wlClient, () => {
> +    logger.info(`setClientDestroyedCallback`)
>      for (const destroyListener of nativeClientSession.destroyListeners) {
>        destroyListener()
>      }
> -    nativeAppContext.sendClientConnectionsDisconnect()
> +    //nativeAppContext.sendClientConnectionsDisconnect()
>      nativeClientSession.destroyListeners = []
>      nativeClientSession.destroyed = true
>    })
> ```

I'm curious how this fix changes the behaviour as I've noticed it doesn't change anything when testing myself. `nativeAppContext.sendClientConnectionsDisconnect()` tells the browser to close all websocket connections that are used to forward display protocol (and other stuff) to the browser, but only *after* the corresponding wayland client is closed. 🤔

**Reply by J-Bu** [reply]
Maybe it just changed the timing, before that change it never worked for me but after it it only worked sometimes like you described is the case for you without the change.

My initial tests where on a VMware ESX hosted virtual machine without hw acceleration which might also be quite slow compared to other real systems.

**Reply by Zubnix** [reply]
I've been doing some experimentation and although I can´t reliably reproduce it, I do think there is a bug in the XWayland websocket management when a remote proxy is restarted while the browser session is not. The browser session fruitlessly tries to reconnect to the restarted remote proxy but fails because there is no more XWayland state associated with those connections.

I'm not sure if this is the cause of the issue but I'm going to fix this either way.

**Reply by J-Bu** [reply]
Thanks for your effort I really appreciate it. If I can help in any way please let me know

**Reply by Zubnix** [reply]
If you're feeling adventurous, you could try https://github.com/udevbe/greenfield/tree/fix-frame-callback It has some fixes and it seems to -at least for me- improve things, although not 100%.

**Reply by J-Bu** [reply, marked as answer]
Sorry for the late response time I finally found some time to look into this again. It is working much better now with the current master. Thank you
